### PR TITLE
Fused attention instances & padding tests

### LINKED ISSUE #391: Not expose Acc0ElementOp compute detail to device level -- Originated by Anthony.
We should never require user to specify `Acc0ElementOp = ResetNaNToMinusInfinity` because users are not concerned with how we handle padding with NaNs and INFs. What I mean is to keep `Acc0ElementOp = scale` as it is, and leave ResetNaNToMinusInfinity details to gridwise level. To reiterate: we can and should do ResetNaNToMinusInfinity somewhere somehow once N padding is on.

And once we've done that, the difference between the two example code should be pretty minimal. We may consider reorg the code into what @poyenc did with https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/example/01_gemm/run_gemm_example.inc

_Originally posted by @rosenrodt in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/385#discussion_r956874214_

### PR DESCRIPTION
**Pull request by rosenrodt**
- Added padding tests
- Added more attention kernel instances; showing improvements in mainline 9110 and even more so in 10768 compiler (kernels having compiler issue in 10768 compiler are not added)
- Streamlined usage where `ResetNaNToMinusInf` is now hidden from user facing device op; closes #391 
- Minor changes in attention kernel: use compile-time conditionals to load OOB value as NaN only after padding is enabled so we can still leverage buffer_load intrinsic's OOB trick (automatic zeroing) when padding is not needed
- Disabled KPadding for now since it is unimplemented

New instances improved some sizes over #345 (MI250; hipclang mainline 9110)
```
$ build/bin/test_batched_gemm_softmax_gemm_fp16 --gtest_filter=*Bench* --gtest_also_run_disabled_tests | grep -P 'c_g_m_o|Best'
c_g_m_o: dim 3, lengths {768, 256, 64}, strides {16384, 64, 1}
Best Perf: 0.22624 ms, 56.9523 TFlops, 444.94 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 64, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 256, 128}, strides {32768, 128, 1}
Best Perf: 0.449394 ms, 57.3435 TFlops, 447.996 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 64, 256, 32, 8, 8, 64, 128, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 512, 64}, strides {32768, 64, 1}
Best Perf: 0.755331 ms, 68.2345 TFlops, 266.541 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 64, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 512, 128}, strides {65536, 128, 1}
Best Perf: 1.47553 ms, 69.8591 TFlops, 272.887 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 64, 256, 32, 8, 8, 64, 128, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 1024, 64}, strides {65536, 64, 1}
^K^BBest Perf: 2.82383 ms, 73.0068 TFlops, 142.591 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 64, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 1024, 128}, strides {131072, 128, 1}
Best Perf: 5.04522 ms, 81.7243 TFlops, 159.618 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 128, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 2048, 64}, strides {131072, 64, 1}
Best Perf: 10.9431 ms, 75.3562 TFlops, 73.5901 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 64, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 2048, 128}, strides {262144, 128, 1}
Best Perf: 19.0034 ms, 86.788 TFlops, 84.7539 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 128, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 4096, 64}, strides {262144, 64, 1}
Best Perf: 43.2621 ms, 76.2454 TFlops, 37.2292 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 64, 32, 2, Default>
c_g_m_o: dim 3, lengths {768, 4096, 128}, strides {524288, 128, 1}
Best Perf: 74.3064 ms, 88.7819 TFlops, 43.3506 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 32, 8, 8, 128, 128, 32, 2, Default>
```

Padding test results:
```
$ build/bin/test_batched_gemm_softmax_gemm_fp16 --gtest_filter=*Adhoc* | grep -P 'c_g_m_o|Best'
c_g_m_o: dim 3, lengths {24, 49, 64}, strides {3136, 64, 1}
Best Perf: 0.027328 ms, 0.539803 TFlops, 22.0328 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 64, 32, 8, 8, 128, 128, 32, 2, MNOPadding>
c_g_m_o: dim 3, lengths {24, 64, 64}, strides {4096, 64, 1}
Best Perf: 0.0240161 ms, 0.802278 TFlops, 28.9086 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 64, 32, 8, 8, 128, 128, 32, 2, MNOPadding>
c_g_m_o: dim 3, lengths {24, 1020, 128}, strides {130560, 128, 1}
Best Perf: 0.203889 ms, 47.0273 TFlops, 92.2103 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 64, 8, 8, 128, 128, 32, 2, MNOPadding>
c_g_m_o: dim 3, lengths {24, 576, 64}, strides {36864, 64, 1}
Best Perf: 0.126496 ms, 16.1145 TFlops, 55.9533 GB/s, DeviceBatchedGemmSoftmaxGemm_Xdl_CShuffle<256, 128, 128, 64, 8, 8, 128, 128, 32, 2, MNOPadding>
```

### CODE REVIEW

**Comment by rosenrodt** [comment]
The 9110 compiler seems to be showing its limits. Had to work round / disable a lot more instances.

File: library/src/tensor_operation_instance/gpu/batched_gemm_gemm/device_batched_gemm_gemm_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp [hunk 1]
```diff
@@ -35,11 +35,21 @@ using device_batched_gemm_gemm_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_inst
...
+        // DeviceBatchedGemmGemm_Xdl_CShuffle<     Row,      Col,      Row,     Row,   F16,    F16,    F16,   F16,     F32,      F16, PassThrough, PassThrough, PassThrough, PassThrough, PassThrough,    GemmDefault,        1,   256,    256,   128,    32,    64,    32,   8,   8,    2,   32,   32,     2,     4,     2,     S<4, 64, 1>,     S<1, 0, 2>,     S<1, 0, 2>,              2,              8,              8,      true,      S<4, 64, 1>,      S<1, 0, 2>,      S<1, 0, 2>,               2,               8,               8,       true,     S<16, 16, 1>,      S<0, 2, 1>,      S<0, 2, 1>,               1,               4,               2,      false,           1,           2,               S<1, 32, 1, 8>,               8>, // failed validation on MI100
```
Comment (asroy) [inline library/src/tensor_operation_instance/gpu/batched_gemm_gemm/device_batched_gemm_gemm_xdl_cshuffle_f16_f16_f16_f16_gmk_gnk_gno_gmo_instance.cpp:39]: [Future PR] define macro guard ``CK_WORKAROUND_SWDEV_XXX``` in ```ck.hpp``` and use it to switch off these instances

File: test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp [hunk 2]
```diff
@@ -37,3 +104,58 @@ TYPED_TEST(TestBatchedGemmSoftmaxGemmFP16, DISABLED_Bench_FP16)
...
+{
+    int P = 120; // requires padding
+    int Q = 128; // do not require padding
+
+    // IsSupported(M, N, K, O)
+    // clang-format off
+    EXPECT_TRUE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::Default>{}.IsSupported(Q, Q, Q, Q));
+    EXPECT_TRUE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::MPadding>{}.IsSupported(P, Q, Q, Q));
+    EXPECT_TRUE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::NPadding>{}.IsSupported(Q, P, Q, Q));
+    // EXPECT_TRUE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::KPadding>{}.IsSupported(Q, Q, P, Q));
```
Comment (asroy) [inline test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp:121]: are these commented-out Still needed?

File: test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp [hunk 3]
```diff
@@ -37,3 +104,58 @@ TYPED_TEST(TestBatchedGemmSoftmaxGemmFP16, DISABLED_Bench_FP16)
...
+    // EXPECT_TRUE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::MNKOPadding>{}.IsSupported(P, P, P, P));
+    // clang-format on
+}
+
+TEST(TestBatchedGemmSoftmaxGemmInterface, GemmSpecializationSizeMismatch)
+{
+    // IsSupported(M, N, K, O)
+    // clang-format off
+    EXPECT_FALSE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::Default>{}.IsSupported(128, 128, 120, 128));
+    // EXPECT_FALSE(DeviceInstanceWrapper_TNTT_FP16_M128_N128_K32_O128<GemmSpecialization::MNKPadding>{}.IsSupported(128, 128, 128, 120));
```
Comment (asroy) [inline test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp:142]: are these commented-out Still needed?

**Review by asroy — APPROVED** [review]

File: test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp (hunk 2 above)
Comment (rosenrodt) [reply test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp:121]: KPadding is not supported by attention kernels yet. We still haven't figured out how to reset NaN to 0 for padded K indices, while at the same time resetting NaN to -inf for padded N indices

File: test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp (hunk 3 above)
Comment (rosenrodt) [reply test/batched_gemm_softmax_gemm/test_batched_gemm_softmax_gemm_fp16.cpp:142]: Anything that touches KPadding is commented out for now

**Pull request merged**